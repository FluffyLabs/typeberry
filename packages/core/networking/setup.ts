import QuicLogger, { formatting, LogLevel, StreamHandler } from "@matrixai/logger";
import { events, QUICClient, type QUICConnection, QUICServer, QUICSocket } from "@matrixai/quic";
import { BytesBlob } from "@typeberry/bytes";
import type { Ed25519Pair } from "@typeberry/crypto/ed25519.js";
import { Level, Logger } from "@typeberry/logger";
import {
  altNameRaw,
  certToPEM,
  ed25519AsJsonWebKeyPair,
  generateCertificate,
  type PeerInfo,
  privateKeyToPEM,
} from "./certificate.js";
import { getQuicClientCrypto, getQuicServerCrypto } from "./crypto.js";
import type { DialOptions } from "./network.js";
import { peerVerification } from "./peer-verification.js";
import { type PeerAddress, PeersManagement } from "./peers.js";
import { QuicNetwork } from "./quic-network.js";
import { QuicPeer } from "./quic-peer.js";
import { addEventListener } from "./quic-utils.js";

const logger = Logger.new(import.meta.filename, "net");

/** Networking server part options. */
export type Options = {
  /** Peer's ed25519key. */
  key: Ed25519Pair;
  /** Host interface to bound to. */
  host: string;
  /** Port to listen on. Use `0` for random. */
  port: number;
  /** Supported ALPNs (protocols) both for the inbound and outbound connections. */
  protocols: string[];
};

enum CloseReason {
  PeerIdMismatch = 0,
  DuplicateConnection = 1,
  ConnectionFromOurself = 2,
}

export class Quic {
  /** Setup QUIC socket and start listening for connections. */
  static async setup({ host, port, protocols, key }: Options): Promise<QuicNetwork> {
    const quicLoggerLvl = logger.getLevel() > Level.TRACE ? LogLevel.WARN : LogLevel.DEBUG;
    const quicLogger = new QuicLogger("quic", quicLoggerLvl, [
      new StreamHandler(formatting.format`${formatting.level}:${formatting.keys}:${formatting.msg}`),
    ]);

    // Load keypair
    const keyPair = ed25519AsJsonWebKeyPair(key);
    const privKeyPEM = await privateKeyToPEM(keyPair);
    const cert = await generateCertificate({
      certId: BytesBlob.blobFromString("QUIC Networking"),
      subjectKeyPair: keyPair,
      issuerKeyPair: keyPair,
    });

    const lastConnectedPeer = peerVerification();

    // QUICConfig
    const config = {
      keepAliveIntervalTime: 3000,
      maxIdleTimeout: 6000,
      applicationProtos: protocols,
      cert: certToPEM(cert),
      key: privKeyPEM,
      verifyPeer: true,
      verifyCallback: lastConnectedPeer.verifyCallback,
    };

    logger.info`🆔 Peer id: ** ${altNameRaw(key.pubKey)}@${host}:${port} ** (pubkey: ${key.pubKey})`;
    // Shared injected UDP socket
    const socket = new QUICSocket({
      logger: quicLogger.getChild("socket"),
    });

    // Start server on the socket.
    const server = new QUICServer({
      socket,
      config,
      crypto: getQuicServerCrypto(key),
      logger: quicLogger.getChild("server"),
    });

    // peer management
    const peers = new PeersManagement<QuicPeer>();

    // basic error handling
    addEventListener(server, events.EventQUICServerError, (error) => logger.error`🛜  Server error: ${error}`);
    addEventListener(server, events.EventQUICServerClose, (ev) => logger.error`🛜  Server stopped: ${ev}`);

    // handling incoming session
    addEventListener(server, events.EventQUICServerConnection, async (ev) => {
      const conn = ev.detail;
      if (lastConnectedPeer.info === null) {
        await conn.stop();
        return;
      }

      if (lastConnectedPeer.info.key.isEqualTo(key.pubKey)) {
        logger.log`🛜 Rejecting connection from ourself from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.ConnectionFromOurself });
        return;
      }

      if (peers.isConnected(lastConnectedPeer.info.id)) {
        logger.log`🛜 Rejecting duplicate connection with peer ${lastConnectedPeer.info.id} from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.DuplicateConnection });
        return;
      }

      logger.log`🛜 Server handshake with ${conn.remoteHost}:${conn.remotePort}`;
      newPeer(conn, lastConnectedPeer.info);
      lastConnectedPeer.info = null;
      await conn.start();
    });

    // connecting to a peer
    async function dial(peer: PeerAddress, options: DialOptions): Promise<QuicPeer> {
      return doDial();

      async function doDial() {
        const peerDetails = peerVerification();
        const clientLater = QUICClient.createQUICClient(
          {
            socket: socket,
            host: peer.host,
            port: peer.port,
            crypto: getQuicClientCrypto(),
            config: {
              ...config,
              verifyCallback: peerDetails.verifyCallback,
            },
            logger: quicLogger.getChild("client"),
          },
          {
            signal: options.signal,
          },
        );
        const client = await clientLater;

        if (peerDetails.info === null) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error("Client connected, but there is no peer details!");
        }

        if (options.verifyName !== undefined && options.verifyName !== peerDetails.info.id) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error(
            `Client connected, but the id didn't match. Expected: ${options.verifyName}, got: ${peerDetails.info.id}`,
          );
        }

        addEventListener(client, events.EventQUICClientClose, () => {
          logger.log`⚰️ Client connection closed.`;
        });

        addEventListener(client, events.EventQUICClientError, (error) => {
          logger.error`🔴 Client error: ${error.detail}`;
        });

        logger.log`🤝 Client handshake with: ${peer.host}:${peer.port}`;
        return newPeer(client.connection, peerDetails.info);
      }
    }

    function newPeer(conn: QUICConnection, peerInfo: PeerInfo) {
      const peer = new QuicPeer(conn, peerInfo);
      addEventListener(peer.conn, events.EventQUICConnectionClose, () => peers.peerDisconnected(peer));
      peers.peerConnected(peer);
      return peer;
    }

    return new QuicNetwork(socket, server, dial, peers, { host, port });
  }
}
