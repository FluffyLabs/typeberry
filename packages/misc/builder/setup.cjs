const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

const isRelease = process.env.IS_RELEASE !== undefined;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const DIST = `${__dirname}/../../../dist/${outDir}`;
const inputFile = `${__dirname}/pkg.ts`;

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(inputFile, data);

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: isRelease
      ? originalPackageJson.version
      : `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
    sideEffects: false,
    exports: {
      ".": {
        import: "./index.js",
        require: "./index.cjs",
      },
    },
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `
bin/**
packages/**
`,
);

module.exports = {
  inputFile,
  esmOutFile: `${DIST}/index.js`,
  cjsOutFile: `${DIST}/index.cjs`,
  typesInput: `${DIST}/packages/misc/builder/pkg.d.ts`,
};
