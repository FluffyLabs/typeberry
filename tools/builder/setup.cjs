const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(`${__dirname}/pkg.ts`, data);

const DIST = `${__dirname}/../../dist/${outDir}`;

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `tools/**
packages/**
`,
);

module.exports = {
  outFile: `${DIST}/index.js`,
  typesInput: `${DIST}/tools/builder/pkg.d.ts`,
};
