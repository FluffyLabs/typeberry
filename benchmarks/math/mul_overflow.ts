import { suite, add, cycle, complete, configure, save } from '@typeberry/benchmark/setup';

const a = 0xffffffff12;
const b = 0x1234123412;

module.exports = () => suite(
  "Wrapping Multiplication",

  add('multiply and bring back to u32', () => {
    const a1 = a >>> 0;
    const b1 = b >>> 0;
    const c1 = (a1 * b1) >>> 0;
    return c1;
  }),

  add('multiply and take modulus', () => {
    const c1 = (a * b) % 2**32;
    return c1;
  }),

  cycle(),
  complete(),
  configure({}),
  ...save(__filename),
)
