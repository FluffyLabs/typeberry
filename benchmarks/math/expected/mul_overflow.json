{
  "name": "Wrapping Multiplication",
  "date": "2024-09-02T04:51:58.690Z",
  "version": null,
  "results": [
    {
      "name": "multiply and bring back to u32",
      "ops": 145369026.23,
      "margin": 5.03,
      "percentSlower": 0
    },
    {
      "name": "multiply and take modulus",
      "ops": 141065889.97,
      "margin": 5.62,
      "percentSlower": 2.96
    }
  ],
  "fastest": {
    "name": "multiply and bring back to u32",
    "index": 0
  },
  "slowest": {
    "name": "multiply and take modulus",
    "index": 1
  }
}